### ISSUE TITLE
Can't use Prettier within a browser

### ISSUE THREAD

**Issue by bvaughn** [prettier/prettier#3296]
Following up from [this Twitter comment](https://twitter.com/suchipi/status/932517379367608320); relates to babel/website/issues/1388 and babel/website/pull/1466.

The Babel REPL has been using Prettier via [bundle.run](https://bundle.run/) for some time. Version 1.7.0 broke our ability to do this with the error: "_cosmiconfig requires at least version 4 of Node_". Version 1.8.0 introduced a different error: "_os$2.homedir is not a function_".

It would be ideal if Prettier output a browser-friendly UMD bundle for use-cases like this, but in lieu of that maybe it could at least work via a solution like rundle.run? 😄 

For the time being, we've pinned Babel to use Prettier version 1.6.1. You can try it out [here](https://deploy-preview-1466--babel.netlify.com/repl/#?babili=false&browsers=&build=&builtIns=false&code_lz=MYewdgzgLgBAhjAvDA3gKBjARgLhgcmHwBoMYATPAIgFMrSBfIA&debug=false&circleciRepo=&evaluate=false&lineWrap=false&presets=react%2Cstage-2&prettier=true&targets=&version=6.26.0) if you'd like.

**Comment by mitermayer** [comment]
This sounds reasonable to me, I can try looking into this during the weekend, unless @azz , @suchipi @ikatyang what’s to look into this first

**Comment by bvaughn** [comment]
A UMD bundle would be so-so-so great ❤️  If there's support for that idea, I'd also be happy to help (although I am not familiar with Prettier's codebase)

**Comment by suchipi** [comment]
I think we can fix "os$2.homedir is not a function" with the same approach as in this PR (adding to the rollup-plugin-replace config): https://github.com/prettier/prettier/pull/2930

"cosmiconfig requires at least version 4 of Node" might require a change to cosmiconfig.

**Comment by suchipi** [comment]
Also, since we are already using rollup, we are probably not too far from generating a working UMD bundle.

**Comment by ikatyang** [comment]
How about extracting `prettier.format()` to another file? The browser version seems no need to read config file (`fs`-related), it only cares about `prettier.format()`.

**Comment by bvaughn** [comment]
A trimmed down version for the browser would be slick.

Would also be nice if there were multiple UMD targets - so we could avoid loading parsers for targets/languages we don't care about (eg CSS).

But I don't want to be greedy. 😁

**Comment by duailibe** [comment]
The parsers are already bundled separately! Seems like we could do is have another bundle for the format function and we’ll have all the pieces in place, just needing some work so they can work together in the browser

**Comment by azz** [comment]
It already kind of works - the files [here](https://github.com/prettier/prettier/tree/master/website/static/lib) are built with rollup and already run in the browser, you just need this gross patch:

https://github.com/prettier/prettier/blob/b4e7942bc080d5e4930527914bdd8fd43f9e3180/website/static/worker.js#L5-L26

We can definitely improve this, splitting the cli into a separate bundle is a great idea. I like the idea of making a UMD bundle, too.

**Comment by duailibe** [comment]
Now that I've worked a bit more on the playground there are a few problemas we need to sort to get this working:

- native node modules (see how we shim on the playground https://github.com/prettier/prettier/issues/3296#issuecomment-345992717)
- lazy load parsers (in the playground we patch `require`).

Does it make sense to ship a bundled version with those shims and our custom `require` ?

**Comment by suchipi** [comment]
I think we should ship the shims, but instead of doing the lazy require stuff or using a custom require, we should ship a UMD bundle containing a factory function that users plug the parsers into, and we should ship the parser versions we vendor as UMD bundles.

**Comment by suchipi** [comment]
So eventual CommonJS usage would be something like:
```js
const makePrettier = require("prettier/bundle/makePrettier"); // or whatever
const flowParser = require("prettier/bundle/flow");
const prettier = makePrettier({ flowParser })
```

And you would only ship the parsers you need

**Comment by siefkenj** [comment]
For my usage, it would be nice to be able to export prettier without a parser. I'm currently developing a parser out-of-tree, and I need to use the printDoc functions but not necessarily any parsing functions.

**Comment by j-f1** [comment]
@siefkenj Check out #3511 and #3536.

**Comment by duailibe** [comment]
I'm gonna start working on this now :)

**Comment by azz** [comment]
@duailibe Wow, I was just typing up #3574.

**Comment by duailibe** [comment]
Can't work on this right now, if anyone wants to take a shot at it (cc @azz)

**Comment by Huvik** [comment]
Any news on this? :) I think a lot of tools could benefit from it for example Graphql playground https://github.com/graphcool/graphql-playground/issues/327

**Comment by nickmccurdy** [comment]
This would be great for using Prettier in browser extensions! See https://github.com/krawaller/prettier-js-extension.

**Comment by Huvik** [comment]
Thx we are rocking it in playground now :) https://github.com/prismagraphql/graphql-playground/pull/702

**Comment by azz** [comment]
Yep, shipped in 1.13.0:
https://prettier.io/docs/en/browser.html
